File: snippets/go/v0_11/concurrency/customer_id.go
package concurrency

import (
	"context"

	"github.com/inngest/inngestgo"
	"github.com/inngest/inngestgo/step"
)

type Bucket struct {
	Fetch func(uri string) (any, error)
}

func parseFile(any) {

}

func CustomerIDSnippet() {
	bucket := Bucket{}
	client, _ := inngestgo.NewClient(inngestgo.ClientOpts{AppID: "my-app"})

	// !snippet:start
	type EventData struct {
		CustomerID string `json:"customer_id"`
		FileURI    string `json:"file_uri"`
	}

	_, err := inngestgo.CreateFunction(
		client,
		inngestgo.FunctionOpts{
			Name: "Process customer csv import",
			ID:   "process-customer-csv-import",
			Concurrency: []inngestgo.ConfigStepConcurrency{
				{
					Limit: 1,
					Key:   inngestgo.StrPtr("event.data.customer_id"), // You can use any piece of data from the event payload
				},
			},
		},
		inngestgo.EventTrigger("csv/file.uploaded", nil),
		func(ctx context.Context, input inngestgo.Input[EventData]) (any, error) {
			_, err := step.Run(ctx, "process-file", func(ctx context.Context) (any, error) {
				file, err := bucket.Fetch(input.Event.Data.FileURI)
				if err != nil {
					return nil, err
				}
				parseFile(file)
				return nil, nil
			})
			if err != nil {
				return err, nil
			}

			return nil, nil
		},
	)
	// !snippet:end

	_ = err
}

File: snippets/py/v0_5/concurrency/customer_id.py
import inngest

inngest_client = inngest.Inngest(app_id="my-app")

class Bucket:
    def fetch(self, uri: str) -> object:
        pass

bucket = Bucket()

def parse_file(file: object):
    pass



# !snippet:start
@inngest_client.create_function(
    fn_id="process-customer-csv-import",
    name="Process customer csv import",
    trigger=inngest.TriggerEvent(event="csv/file.uploaded"),
    concurrency=[
        inngest.Concurrency(
            limit=1,
            key="event.data.customerId"  # You can use any piece of data from the event payload
        )
    ]
)
async def process_csv_import(ctx: inngest.Context):
    async def process_file():
        file = await bucket.fetch(ctx.event.data.file_uri)
        parse_file(file)

    await ctx.step.run("process-file", process_file)
    return {"message": "success"}
